File: csharp/demo/demo13a-mutex.cs
/*
 * MUTEXES
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Demo13A : IRunnable
{
    private Mutex mut;
    private int counter;


    public void run()
    {
        const int NUM_THREADS = 16;

        mut = new Mutex();
        counter = 0;

        var lstTh = new List<Thread>();

        for (int i = 0; i < NUM_THREADS; ++i)
        {
            lstTh.Add(new Thread(doTask));
        }

        lstTh.ForEach(th => th.Start());
        lstTh.ForEach(th => th.Join());

        mut.Dispose();

        Console.WriteLine("counter = " + counter);
        // We are sure that counter = 16000
    }


    private void doTask()
    {
        Thread.Sleep(1000);

        mut.WaitOne();

        for (int i = 0; i < 1000; ++i)
            ++counter;

        mut.ReleaseMutex();
    }
}

File: cpp-std/demo13a-mutex.cpp
/*
MUTEXES

Without synchronization (by a mutex), we are not sure that result = 30000.
*/


#include <iostream>
#include <chrono>
#include <thread>
#include <mutex>
using namespace std;



std::mutex mut;
int counter = 0;



void doTask() {
    std::this_thread::sleep_for(std::chrono::seconds(1));

    mut.lock();

    for (int i = 0; i < 10000; ++i)
        ++counter;

    mut.unlock();
}



int main() {
    constexpr int NUM_THREADS = 3;
    std::thread lstTh[NUM_THREADS];


    for (auto&& th : lstTh) {
        th = std::thread(doTask);
    }


    for (auto&& th : lstTh) {
        th.join();
    }


    cout << "counter = " << counter << endl;
    return 0;
}
